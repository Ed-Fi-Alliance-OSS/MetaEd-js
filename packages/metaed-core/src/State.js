// @flow
import type { InputDirectory } from './task/FileSystemFilenameLoader';
import type { FileSet } from './task/MetaEdFile';
import type { FileIndex } from './task/FileIndex';
import type { MetaEdGrammar } from './grammar/gen/MetaEdGrammar';
import type { ValidationFailure } from './validator/ValidationFailure';
import type { MetaEdEnvironment } from './MetaEdEnvironment';
import { newMetaEdEnvironment } from './MetaEdEnvironment';
import type { PluginManifest } from './plugin/PluginTypes';

export type State = {
  // the collection of error messages from syntax and semantic validation, and other processes
  validationFailure: Array<ValidationFailure>,

  // the specified directories to load .metaed files from
  inputDirectories: ?InputDirectory[],

  // filepaths to exclude from loading, usually used to allow upstream tasks to provide their own versions in a FileSet
  // e.g. files that are open and unsaved in an editor
  filepathsToExclude: Set<string>,

  // the set of files whose contents have been loaded
  loadedFileSet: Array<FileSet>,

  // the line number indexing for the concatenation of all .metaed files
  fileIndex: ?FileIndex,

  // the ANTLR parse tree of the concatenated .metaed files
  parseTree: ?MetaEdGrammar,

  // the MetaEd environment
  metaEd: MetaEdEnvironment,

  // the directory to scan for plugins
  pluginScanDirectory: ?string,

  // plugins
  pluginManifest: Array<PluginManifest>,
};

export const newState: () => State = () =>
  ({
    validationFailure: [],
    inputDirectories: [],
    filepathsToExclude: new Set(),
    loadedFileSet: [],
    fileIndex: null,
    parseTree: null,
    metaEd: newMetaEdEnvironment(),
    pluginScanDirectory: null,
    pluginManifest: [],
  });
