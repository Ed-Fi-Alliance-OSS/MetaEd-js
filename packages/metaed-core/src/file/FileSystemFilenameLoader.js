// @flow
import ffs from 'final-fs';
import path from 'path';
import winston from 'winston';
import { createMetaEdFile } from './MetaEdFile';
import type { FileSet } from './MetaEdFile';
import type { State } from '../State';
import type { MetaEdConfiguration } from '../MetaEdConfiguration';

export function loadFiles(state: State): boolean {
  let success: boolean = true;
  const { metaEdConfiguration }: { metaEdConfiguration: MetaEdConfiguration } = state;

  if (metaEdConfiguration.projects.length !== metaEdConfiguration.projectPaths.length) {
    winston.error('FileSystemFilenameLoader: project metadata must be same length as project paths');
    return false;
  }

  if (!Array.isArray(state.inputDirectories)) state.inputDirectories = [];

  for (let i = 0; i < metaEdConfiguration.projects.length; i += 1) {
    const projectExtension =
      metaEdConfiguration.projects[i].projectExtension ||
      (metaEdConfiguration.projects[i].namespaceName === 'edfi' ? '' : 'EXTENSION');

    state.inputDirectories.push({
      path: metaEdConfiguration.projectPaths[i],
      namespaceName: metaEdConfiguration.projects[i].namespaceName,
      projectExtension,
      projectName: metaEdConfiguration.projects[i].projectName,
      isExtension: metaEdConfiguration.projects[i].namespaceName !== 'edfi',
    });
  }

  const fileSets: FileSet[] = [];
  let filenamesFoundInDirectories: boolean = false;

  state.inputDirectories.forEach(inputDirectory => {
    const fileSet: FileSet = {
      namespaceName: inputDirectory.namespaceName,
      projectExtension: inputDirectory.projectExtension,
      projectName: inputDirectory.projectName,
      isExtension: inputDirectory.isExtension,
      files: [],
    };

    try {
      const filenames: string[] = ffs
        .readdirRecursiveSync(inputDirectory.path, true, inputDirectory.path)
        .filter(filename => filename.endsWith('.metaed'));

      if (filenames.length > 0) {
        filenamesFoundInDirectories = true;
      }

      const filenamesToLoad: string[] = filenames.filter(filename => !state.filePathsToExclude.has(filename));

      filenamesToLoad.forEach(filename => {
        const contents = ffs.readFileSync(filename, 'utf-8');
        const metaEdFile = createMetaEdFile(path.dirname(filename), path.basename(filename), contents);
        fileSet.files.push(metaEdFile);
      });

      winston.info(
        `  ${inputDirectory.path} (${filenamesToLoad.length} .metaed file${filenamesToLoad.length > 1 ? 's' : ''} loaded)`,
      );

      fileSets.push(fileSet);
    } catch (exception) {
      winston.error(`FileSystemFilenameLoader: Unable to read files at location '${inputDirectory.path}'.`, exception);
      if (exception.code === 'ENOTEMPTY' || exception.code === 'EPERM') {
        winston.error('Please close any files or folders that may be open in other applications.');
      }
      success = false;
    }
  });

  if (!filenamesFoundInDirectories) {
    winston.error('No MetaEd files found in any input directory.');
    success = false;
  }

  state.loadedFileSet.push(...fileSets);
  return success;
}
