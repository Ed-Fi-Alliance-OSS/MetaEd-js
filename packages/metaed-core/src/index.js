// @flow

// Functions
export {
  deepFreeze,
  deepFreezeAssign,
  prependIndefiniteArticle,
  orderByProp,
  versionSatisfies,
  V2Only,
  V3OrGreater,
} from './Utility';
export { executePipeline } from './task/Pipeline';
export { createMetaEdFile } from './task/MetaEdFile';
export { validateConfiguration } from './task/ValidateConfiguration';
export { loadCoreBufferedFiles, loadExtensionBufferedFiles } from './task/BufferFileLoader';
export { loadFiles } from './task/FileSystemFilenameLoader';
export { loadPlugins, scanForPlugins } from './task/LoadPlugins';
export { newState } from './State';
export { newPipelineOptions } from './task/PipelineOptions';
export { loadFileIndex } from './task/LoadFileIndex';
export { buildParseTree } from './task/BuildParseTree';
export { buildMetaEd } from './grammar/ParseTreeBuilder';
export { execute as walkBuilders } from './task/WalkBuilders';
export { execute as runValidators } from './task/RunValidators';
export { execute as runEnhancers } from './task/RunEnhancers';
export { execute as runGenerators } from './task/RunGenerators';
export { fileMapForFailure } from './task/FileMapForFailure';
export {
  addEntity,
  getAllEntities,
  getAllTopLevelEntities,
  getEntity,
  getTopLevelCoreEntity,
  getEntitiesOfType,
  getAllEntitiesNoSimpleTypes,
} from './model/EntityRepository';
export { addProperty, getAllProperties, getPropertiesOfType, newPropertyIndex } from './model/property/PropertyRepository';
export { isReferenceProperty, isSharedProperty, allPropertyTypes } from './model/property/PropertyType';
export { allEntityModelTypes, allTopLevelEntityModelTypes, topLevelCoreEntityModelTypes } from './model/ModelType';

// Builders
export { AssociationBuilder } from './builder/AssociationBuilder';
export { AssociationExtensionBuilder } from './builder/AssociationExtensionBuilder';
export { AssociationSubclassBuilder } from './builder/AssociationSubclassBuilder';
export { ChoiceBuilder } from './builder/ChoiceBuilder';
export { CommonBuilder } from './builder/CommonBuilder';
export { CommonExtensionBuilder } from './builder/CommonExtensionBuilder';
export { DecimalTypeBuilder } from './builder/DecimalTypeBuilder';
export { DescriptorBuilder } from './builder/DescriptorBuilder';
export { DomainBuilder } from './builder/DomainBuilder';
export { DomainEntityBuilder } from './builder/DomainEntityBuilder';
export { DomainEntityExtensionBuilder } from './builder/DomainEntityExtensionBuilder';
export { DomainEntitySubclassBuilder } from './builder/DomainEntitySubclassBuilder';
export { EnumerationBuilder } from './builder/EnumerationBuilder';
export { IntegerTypeBuilder } from './builder/IntegerTypeBuilder';
export { InterchangeBuilder } from './builder/InterchangeBuilder';
export { NamespaceInfoBuilder } from './builder/NamespaceInfoBuilder';
export { SharedDecimalBuilder } from './builder/SharedDecimalBuilder';
export { SharedIntegerBuilder } from './builder/SharedIntegerBuilder';
export { SharedStringBuilder } from './builder/SharedStringBuilder';
export { StringTypeBuilder } from './builder/StringTypeBuilder';

// Flow types
export type { State } from './State';
export type { InputDirectory } from './task/FileSystemFilenameLoader';
export type { FileSet, MetaEdFile } from './task/MetaEdFile';
export type { FileIndex } from './task/FileIndex';
export type { PipelineOptions } from './task/PipelineOptions';
export type { Validator } from './validator/Validator';
export type { ValidationFailure } from './validator/ValidationFailure';
export type { Enhancer } from './enhancer/Enhancer';
export type { EnhancerResult } from './enhancer/EnhancerResult';
export type { Generator } from './generator/Generator';
export type { GeneratorResult } from './generator/GeneratorResult';
export type { GeneratedOutput } from './generator/GeneratedOutput';
export type { MetaEdEnvironment, SemVer } from './MetaEdEnvironment';
export type { PluginEnvironment } from './plugin/PluginEnvironment';
export type { MetaEdConfiguration } from './MetaEdConfiguration';
export type { MetaEdPlugin, PluginManifest } from './plugin/PluginTypes';

// Properties
export type { AssociationProperty, AssociationPropertySourceMap } from './model/property/AssociationProperty';
export type { BooleanProperty, BooleanPropertySourceMap } from './model/property/BooleanProperty';
export type { ChoiceProperty, ChoicePropertySourceMap } from './model/property/ChoiceProperty';
export type { CommonProperty, CommonPropertySourceMap } from './model/property/CommonProperty';
export type { CurrencyProperty, CurrencyPropertySourceMap } from './model/property/CurrencyProperty';
export type { DateProperty, DatePropertySourceMap } from './model/property/DateProperty';
export type { DecimalProperty, DecimalPropertySourceMap } from './model/property/DecimalProperty';
export type { DescriptorProperty, DescriptorPropertySourceMap } from './model/property/DescriptorProperty';
export type { DomainEntityProperty, DomainEntityPropertySourceMap } from './model/property/DomainEntityProperty';
export type { DurationProperty, DurationPropertySourceMap } from './model/property/DurationProperty';
export type { EntityProperty, EntityPropertySourceMap, HasReferencedEntity } from './model/property/EntityProperty';
export type { EnumerationProperty, EnumerationPropertySourceMap } from './model/property/EnumerationProperty';
export type { InlineCommonProperty, InlineCommonPropertySourceMap } from './model/property/InlineCommonProperty';
export type { IntegerProperty, IntegerPropertySourceMap } from './model/property/IntegerProperty';
export type { MergedProperty, MergedPropertySourceMap } from './model/property/MergedProperty';
export type { PercentProperty, PercentPropertySourceMap } from './model/property/PercentProperty';
export type { PropertyIndex } from './model/property/PropertyRepository';
export type { PropertyType } from './model/property/PropertyType';
export type { ReferentialProperty, ReferentialPropertySourceMap } from './model/property/ReferentialProperty';
export type {
  SchoolYearEnumerationProperty,
  SchoolYearEnumerationPropertySourceMap,
} from './model/property/SchoolYearEnumerationProperty';
export type { SharedDecimalProperty } from './model/property/SharedDecimalProperty';
export type { SharedIntegerProperty } from './model/property/SharedIntegerProperty';
export type { SharedShortProperty } from './model/property/SharedShortProperty';
export type { SharedStringProperty } from './model/property/SharedStringProperty';
export type { ShortProperty, ShortPropertySourceMap } from './model/property/ShortProperty';
export type { SimpleProperty, SimplePropertySourceMap } from './model/property/SimpleProperty';
export type { SharedSimple, SharedSimpleSourceMap } from './model/SharedSimple';
export type { StringProperty, StringPropertySourceMap } from './model/property/StringProperty';
export type { TimeProperty, TimePropertySourceMap } from './model/property/TimeProperty';
export type { YearProperty, YearPropertySourceMap } from './model/property/YearProperty';

// Entities
export type { AbstractEntity, AbstractEntitySourceMap } from './model/AbstractEntity';
export type { Association, AssociationSourceMap } from './model/Association';
export type { AssociationExtension, AssociationExtensionSourceMap } from './model/AssociationExtension';
export type { AssociationSubclass, AssociationSubclassSourceMap } from './model/AssociationSubclass';
export type { Choice, ChoiceSourceMap } from './model/Choice';
export type { Common, CommonSourceMap } from './model/Common';
export type { CommonExtension, CommonExtensionSourceMap } from './model/CommonExtension';
export type { DecimalType, DecimalTypeSourceMap } from './model/DecimalType';
export type { Descriptor, DescriptorSourceMap } from './model/Descriptor';
export type { Domain, DomainSourceMap } from './model/Domain';
export type { DomainEntity, DomainEntitySourceMap } from './model/DomainEntity';
export type { DomainEntityExtension, DomainEntityExtensionSourceMap } from './model/DomainEntityExtension';
export type { DomainEntitySubclass, DomainEntitySubclassSourceMap } from './model/DomainEntitySubclass';
export type { DomainItem, DomainItemSourceMap } from './model/DomainItem';
export type { EntityRepository } from './model/EntityRepository';
export type { Enumeration, EnumerationSourceMap } from './model/Enumeration';
export type { EnumerationItem, EnumerationItemSourceMap } from './model/EnumerationItem';
export type { IntegerType, IntegerTypeSourceMap } from './model/IntegerType';
export type { Interchange, InterchangeSourceMap } from './model/Interchange';
export type { InterchangeExtension } from './model/InterchangeExtension';
export type { InterchangeItem, InterchangeItemSourceMap } from './model/InterchangeItem';
export type { MapTypeEnumeration } from './model/MapTypeEnumeration';
export type { ModelBase, ModelBaseSourceMap } from './model/ModelBase';
export type { ModelType } from './model/ModelType';
export type { NamespaceInfo, NamespaceInfoSourceMap } from './model/NamespaceInfo';
export type { SchoolYearEnumeration } from './model/SchoolYearEnumeration';
export type { SharedDecimal, SharedDecimalSourceMap } from './model/SharedDecimal';
export type { SharedInteger, SharedIntegerSourceMap } from './model/SharedInteger';
export type { SharedString, SharedStringSourceMap } from './model/SharedString';
export type { SourceMap } from './model/SourceMap';
export type { StringType, StringTypeSourceMap } from './model/StringType';
export type { Subdomain, SubdomainSourceMap } from './model/Subdomain';
export type { TopLevelEntity, TopLevelEntitySourceMap } from './model/TopLevelEntity';

// Null Objects
export { NoMapTypeEnumeration } from './model/MapTypeEnumeration';
export { NoTopLevelEntity } from './model/TopLevelEntity';
export { NoSharedSimple } from './model/SharedSimple';
export { NoNamespaceInfo } from './model/NamespaceInfo';

// Factories
export { newMetaEdEnvironment } from './MetaEdEnvironment';
export { newMetaEdPlugin } from './plugin/PluginTypes';
export { newPluginEnvironment } from './plugin/PluginEnvironment';
export { newMetaEdConfiguration, newPluginConfiguration, findDataStandardVersions } from './MetaEdConfiguration';

export { asAssociationProperty, newAssociationProperty } from './model/property/AssociationProperty';
export { asBooleanProperty, newBooleanProperty } from './model/property/BooleanProperty';
export { asChoiceProperty, newChoiceProperty } from './model/property/ChoiceProperty';
export { asCommonProperty, newCommonProperty } from './model/property/CommonProperty';
export { asCurrencyProperty, newCurrencyProperty } from './model/property/CurrencyProperty';
export { asDateProperty, newDateProperty } from './model/property/DateProperty';
export { asDecimalProperty, newDecimalProperty } from './model/property/DecimalProperty';
export { asDescriptorProperty, newDescriptorProperty } from './model/property/DescriptorProperty';
export { asDomainEntityProperty, newDomainEntityProperty } from './model/property/DomainEntityProperty';
export { asDurationProperty, newDurationProperty } from './model/property/DurationProperty';
export { asEnumerationProperty, newEnumerationProperty } from './model/property/EnumerationProperty';
export { asInlineCommonProperty, newInlineCommonProperty } from './model/property/InlineCommonProperty';
export { asIntegerProperty, newIntegerProperty } from './model/property/IntegerProperty';
export { asMergedProperty, newMergedProperty } from './model/property/MergedProperty';
export { asPercentProperty, newPercentProperty } from './model/property/PercentProperty';
export { asReferentialProperty, newReferentialProperty, isReferentialProperty } from './model/property/ReferentialProperty';
export {
  asSchoolYearEnumerationProperty,
  newSchoolYearEnumerationProperty,
} from './model/property/SchoolYearEnumerationProperty';
export { asSharedDecimalProperty, newSharedDecimalProperty } from './model/property/SharedDecimalProperty';
export { asSharedIntegerProperty, newSharedIntegerProperty } from './model/property/SharedIntegerProperty';
export { asSharedShortProperty, newSharedShortProperty } from './model/property/SharedShortProperty';
export { asSharedStringProperty, newSharedStringProperty } from './model/property/SharedStringProperty';
export { asShortProperty, newShortProperty } from './model/property/ShortProperty';
export { asSimpleProperty, newSimpleProperty } from './model/property/SimpleProperty';
export { asStringProperty, newStringProperty } from './model/property/StringProperty';
export { asTimeProperty, newTimeProperty } from './model/property/TimeProperty';
export { asYearProperty, newYearProperty } from './model/property/YearProperty';

export { asAbstractEntity, newAbstractEntity } from './model/AbstractEntity';
export { asAssociation, newAssociation } from './model/Association';
export { asAssociationExtension, newAssociationExtension } from './model/AssociationExtension';
export { asAssociationSubclass, newAssociationSubclass } from './model/AssociationSubclass';
export { asChoice, newChoice } from './model/Choice';
export { asCommon, newCommon, asInlineCommon, newInlineCommon } from './model/Common';
export { asCommonExtension, newCommonExtension } from './model/CommonExtension';
export { asDecimalType, newDecimalType } from './model/DecimalType';
export { asDescriptor, newDescriptor } from './model/Descriptor';
export { asDomain, asDomainBase, newDomain } from './model/Domain';
export { asDomainEntity, newDomainEntity } from './model/DomainEntity';
export { asDomainEntityExtension, newDomainEntityExtension } from './model/DomainEntityExtension';
export { asDomainEntitySubclass, newDomainEntitySubclass } from './model/DomainEntitySubclass';
export { asDomainItem, newDomainItem } from './model/DomainItem';
export { asEnumeration, newEnumeration } from './model/Enumeration';
export { asEnumerationItem, newEnumerationItem } from './model/EnumerationItem';
export { asIntegerType, newIntegerType } from './model/IntegerType';
export { asInterchange, newInterchange } from './model/Interchange';
export { asInterchangeExtension, newInterchangeExtension } from './model/InterchangeExtension';
export { asInterchangeItem, newInterchangeItem } from './model/InterchangeItem';
export { asMapTypeEnumeration, newMapTypeEnumeration } from './model/MapTypeEnumeration';
export { asModelType } from './model/ModelType';
export { newNamespaceInfo } from './model/NamespaceInfo';
export { asSchoolYearEnumeration, newSchoolYearEnumeration } from './model/SchoolYearEnumeration';
export { asSharedDecimal, newSharedDecimal } from './model/SharedDecimal';
export { asSharedInteger, newSharedInteger } from './model/SharedInteger';
export { asSharedString, newSharedString } from './model/SharedString';
export { newSourceMap } from './model/SourceMap';
export { asStringType, newStringType } from './model/StringType';
export { asSubdomain, newSubdomain } from './model/Subdomain';
export { asTopLevelEntity, newTopLevelEntity } from './model/TopLevelEntity';

// utilities
export { normalizeDescriptorSuffix, normalizeEnumerationSuffix } from './Utility';

// for plugin testing
export { MetaEdTextBuilder } from './grammar/MetaEdTextBuilder';
