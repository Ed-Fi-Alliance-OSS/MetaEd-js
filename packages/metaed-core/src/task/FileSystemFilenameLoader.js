// @flow
import ffs from 'final-fs';
import path from 'path';
import winston from 'winston';
import { createMetaEdFile } from './MetaEdFile';
import type { FileSet } from './MetaEdFile';
import type { State } from '../State';
import type { MetaEdConfiguration } from '../MetaEdConfiguration';

export type InputDirectory = {
  path: string,
  namespaceName: string,
  projectExtension: string,
  projectName: string,
  isExtension: boolean,
};

export function loadFiles(state: State): boolean {
  let success: boolean = true;
  const metaEdConfiguration: MetaEdConfiguration = state.metaEdConfiguration;

  if (metaEdConfiguration.projects.length !== metaEdConfiguration.projectPaths.length) {
    winston.error('FileSystemFilenameLoader: project metadata must be same length as project paths');
    return false;
  }

  if (!Array.isArray(state.inputDirectories)) state.inputDirectories = [];

  for (let i = 0; i < metaEdConfiguration.projects.length; i += 1) {
    const projectExtension =
      metaEdConfiguration.projects[i].projectExtension ||
      (metaEdConfiguration.projects[i].namespaceName === 'edfi' ? '' : 'EXTENSION');

    state.inputDirectories.push({
      path: metaEdConfiguration.projectPaths[i],
      namespaceName: metaEdConfiguration.projects[i].namespaceName,
      projectExtension,
      projectName: metaEdConfiguration.projects[i].projectName,
      isExtension: metaEdConfiguration.projects[i].namespaceName !== 'edfi',
    });
  }

  const fileSets: FileSet[] = [];
  state.inputDirectories.forEach(inputDirectory => {
    const fileSet: FileSet = {
      namespaceName: inputDirectory.namespaceName,
      projectExtension: inputDirectory.projectExtension,
      projectName: inputDirectory.projectName,
      isExtension: inputDirectory.isExtension,
      files: [],
    };

    const filenames: string[] = ffs
      .readdirRecursiveSync(inputDirectory.path, true, inputDirectory.path)
      .filter(filename => filename.endsWith('.metaed'));

    if (filenames.length === 0) {
      winston.error(`No MetaEd files found in input directory ${inputDirectory.path}.`);
      success = false;
    }

    const filenamesToLoad: string[] = filenames.filter(filename => !state.filePathsToExclude.has(filename));

    filenamesToLoad.forEach(filename => {
      const contents = ffs.readFileSync(filename, 'utf-8');
      const metaEdFile = createMetaEdFile(path.dirname(filename), path.basename(filename), contents);
      fileSet.files.push(metaEdFile);
    });

    winston.info(
      `  ${inputDirectory.path} (${filenamesToLoad.length} .metaed file${filenamesToLoad.length > 1 ? 's' : ''} loaded)`,
    );

    fileSets.push(fileSet);
  });

  state.loadedFileSet.push(...fileSets);
  return success;
}
