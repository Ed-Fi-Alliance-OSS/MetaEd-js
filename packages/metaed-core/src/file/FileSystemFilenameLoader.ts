import fs from 'node:fs';
import klawSync from 'klaw-sync';
import path from 'path';
import winston from 'winston';
import { createMetaEdFile } from './MetaEdFile';
import { FileSet } from './MetaEdFile';
import { State } from '../State';
import { MetaEdConfiguration } from '../MetaEdConfiguration';

export function loadFiles(state: State): boolean {
  let success = true;
  const { metaEdConfiguration }: { metaEdConfiguration: MetaEdConfiguration } = state;

  if (metaEdConfiguration.projects.length !== metaEdConfiguration.projectPaths.length) {
    winston.error('FileSystemFilenameLoader: project metadata must be same length as project paths');
    return false;
  }

  if (!Array.isArray(state.inputDirectories)) state.inputDirectories = [];

  for (let i = 0; i < metaEdConfiguration.projects.length; i += 1) {
    const projectExtension =
      metaEdConfiguration.projects[i].projectExtension ||
      (metaEdConfiguration.projects[i].namespaceName === 'EdFi' ? '' : 'EXTENSION');

    state.inputDirectories.push({
      path: metaEdConfiguration.projectPaths[i],
      namespaceName: metaEdConfiguration.projects[i].namespaceName,
      projectExtension,
      projectName: metaEdConfiguration.projects[i].projectName,
      isExtension: metaEdConfiguration.projects[i].namespaceName !== 'EdFi',
    });
  }

  const fileSets: FileSet[] = [];
  let filenamesFoundInDirectories = false;

  state.inputDirectories.forEach((inputDirectory) => {
    const fileSet: FileSet = {
      namespaceName: inputDirectory.namespaceName,
      projectExtension: inputDirectory.projectExtension,
      projectName: inputDirectory.projectName,
      isExtension: inputDirectory.isExtension,
      files: [],
    };

    try {
      const filenames: string[] = klawSync(inputDirectory.path, {
        filter: (item) => ['.metaed', '.metaEd', '.MetaEd', '.METAED'].includes(path.extname(item.path)),
      }).map((klawObject) => klawObject.path);

      if (filenames.length > 0) {
        filenamesFoundInDirectories = true;
      }

      const filenamesToLoad: string[] = filenames.filter((filename) => !state.filePathsToExclude.has(filename));

      filenamesToLoad.forEach((filename) => {
        const contents = fs.readFileSync(filename, 'utf-8');
        const metaEdFile = createMetaEdFile(path.dirname(filename), path.basename(filename), contents);
        fileSet.files.push(metaEdFile);
      });

      winston.info(
        `- ${inputDirectory.path} (${filenamesToLoad.length} .metaed file${filenamesToLoad.length > 1 ? 's' : ''} loaded)`,
      );

      fileSets.push(fileSet);
    } catch (exception) {
      winston.error(`FileSystemFilenameLoader: Unable to read files at location '${inputDirectory.path}'.`, exception);
      if (exception.code === 'ENOTEMPTY' || exception.code === 'EPERM') {
        winston.error('Please close any files or folders that may be open in other applications.');
      }
      success = false;
    }
  });

  if (!filenamesFoundInDirectories) {
    winston.error('No MetaEd files found in any input directory.');
    success = false;
  }

  state.loadedFileSet.push(...fileSets);
  return success;
}
