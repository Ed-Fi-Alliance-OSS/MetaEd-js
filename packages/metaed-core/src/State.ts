import { newMetaEdConfiguration } from './MetaEdConfiguration';
import { newMetaEdEnvironment } from './MetaEdEnvironment';
import { newPipelineOptions } from './pipeline/PipelineOptions';
import { MetaEdConfiguration } from './MetaEdConfiguration';
import { ValidationFailure } from './validator/ValidationFailure';
import { PipelineFailure } from './pipeline/PipelineFailure';
import { EnhancerResult } from './enhancer/EnhancerResult';
import { GeneratorResult } from './generator/GeneratorResult';
import { InputDirectory } from './file/InputDirectory';
import { FileSet } from './file/MetaEdFile';
import { FileIndex } from './file/FileIndex';
import { PipelineOptions } from './pipeline/PipelineOptions';
import { MetaEdGrammar } from './grammar/gen/MetaEdGrammar';
import { MetaEdEnvironment } from './MetaEdEnvironment';
import { PluginManifest } from './plugin/PluginManifest';

/**
 *
 */
export interface State {
  // the project level configuration loaded from the metaed.json file either located at the root level of a project
  // or referenced by the console's --config argument
  metaEdConfiguration: MetaEdConfiguration;

  // the collection of error messages due to an error somewhere in the pipeline e.g. plugin loading
  pipelineFailure: PipelineFailure[];

  // the collection of error messages from syntax and semantic validation, and other processes
  validationFailure: ValidationFailure[];

  // the collection of enhancer results returned by each enhancer
  enhancerResults: EnhancerResult[];

  // the collection of generator results returned by each generator
  generatorResults: GeneratorResult[];

  // the specified directories to load .metaed files from
  inputDirectories: InputDirectory[];
  // file paths to exclude from loading, usually used to allow upstream tasks to provide their own versions in a FileSet
  // e.g. files that are open and unsaved in an editor
  filePathsToExclude: Set<string>;

  // the set of files whose contents have been loaded
  loadedFileSet: FileSet[];

  // the line number indexing for the concatenation of all .metaed files
  fileIndex: FileIndex | null;

  // the ANTLR parse tree of the concatenated .metaed files
  parseTree: MetaEdGrammar | null;

  // the MetaEd environment
  metaEd: MetaEdEnvironment;

  // the directory where metaed outputs artifacts
  outputDirectory: string | null;

  // the directory to scan for plugins
  pluginScanDirectory: string | null;

  // the plugin manifest information loaded from each plugin's package.json
  pluginManifest: PluginManifest[];

  // options for what pipeline steps should run
  pipelineOptions: PipelineOptions;
}

/**
 *
 */
export const newState: () => State = () => ({
  metaEdConfiguration: newMetaEdConfiguration(),
  pipelineFailure: [],
  validationFailure: [],
  enhancerResults: [],
  generatorResults: [],
  inputDirectories: [],
  filePathsToExclude: new Set(),
  loadedFileSet: [],
  fileIndex: null,
  parseTree: null,
  metaEd: newMetaEdEnvironment(),
  outputDirectory: null,
  pluginScanDirectory: null,
  pluginManifest: [],
  pluginConfiguration: new Map(),
  pipelineOptions: newPipelineOptions(),
});
