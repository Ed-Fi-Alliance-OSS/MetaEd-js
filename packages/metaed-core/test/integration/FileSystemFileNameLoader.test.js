// @flow
import ffs from 'final-fs';
import { MetaEdTextBuilder } from '../MetaEdTextBuilder';
import { newState } from '../../src/State';
import { loadFiles } from '../../src/task/FileSystemFilenameLoader';

describe('When a single file', () => {
  beforeAll(() => {
    const metaEdText = MetaEdTextBuilder.build()
    .withStartDomainEntity('DomainEntity1')
    .withDocumentation('doc1')
    .withStringIdentity('Property1', 'doc2', '100')
    .withEndDomainEntity()
    .toString();

    const domainEntity1 = {
      path: '/fake/dir/Domain Entities/DomainEntity1.metaed',
      content: metaEdText,
    };
    ffs.addMockFile(domainEntity1);
  });

  it('Should load the file contents', () => {
    const state = loadFiles(Object.assign(newState(),
      {
        inputDirectories: [{
          path: '/fake/dir',
          namespace: 'edfi',
          projectExtension: '',
          isExtension: false,
        }],
      }));
    const contents = state.loadedFileSet[0].files[0].contents;
    expect(contents).toMatch(new RegExp('Domain Entity'));
    expect(contents).toMatch(new RegExp('DomainEntity1'));
    expect(contents).toMatch(new RegExp('string'));
    expect(contents).toMatch(new RegExp('Property1'));
  });
});

describe('When multiple files', () => {
  beforeAll(() => {
    const metaEdTextDomainEntity = MetaEdTextBuilder.build()
    .withStartDomainEntity('DomainEntity1')
    .withDocumentation('doc')
    .withStringIdentity('Property1', 'doc', '100')
    .withEndDomainEntity()
    .toString();

    const metaEdTextAssociation = MetaEdTextBuilder.build()
    .withStartAssociation('Association1')
    .withDocumentation('doc')
    .withAssociationDomainEntityProperty('Domain1', 'doc')
    .withAssociationDomainEntityProperty('Domain2', 'doc')
    .withIntegerIdentity('Property2', 'doc', '100')
    .withEndDomainEntity()
    .toString();

    const association1 = {
      path: '/fake/dir/Associations/Association1.metaed',
      content: metaEdTextAssociation,
    };
    const domainEntity1 = {
      path: '/fake/dir/Domain Entities/DomainEntity1.metaed',
      content: metaEdTextDomainEntity,
    };
    ffs.addMockFile(association1);
    ffs.addMockFile(domainEntity1);
  });


  it('Should load the file contents', () => {
    const state = loadFiles(Object.assign(newState(),
      {
        inputDirectories: [{
          path: '/fake',
          namespace: 'edfi',
          projectExtension: '',
          isExtension: false,
        }],
      }));
    const associationContents = state.loadedFileSet[0].files[0].contents;
    expect(associationContents).toMatch(new RegExp('Association'));
    expect(associationContents).toMatch(new RegExp('Domain1'));
    expect(associationContents).toMatch(new RegExp('Domain2'));
    expect(associationContents).toMatch(new RegExp('integer'));
    expect(associationContents).toMatch(new RegExp('Property2'));
    expect(associationContents).toMatchSnapshot();

    const domainEntityContents = state.loadedFileSet[0].files[1].contents;
    expect(domainEntityContents).toMatch(new RegExp('Domain Entity'));
    expect(domainEntityContents).toMatch(new RegExp('DomainEntity1'));
    expect(domainEntityContents).toMatch(new RegExp('string'));
    expect(domainEntityContents).toMatch(new RegExp('Property1'));
    expect(associationContents).toMatchSnapshot();
  });
});

