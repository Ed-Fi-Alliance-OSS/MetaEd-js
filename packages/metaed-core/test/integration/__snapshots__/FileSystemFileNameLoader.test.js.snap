// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When multiple files Should load the file contents 1`] = `
"Association Association1
  documentation
  \\"doc\\"
  domain entity Domain1
    documentation
    \\"doc\\"
  domain entity Domain2
    documentation
    \\"doc\\"
  integer Property2
    documentation
    \\"doc\\"
    is part of identity
    max value 100
"
`;

exports[`When multiple files Should load the file contents 2`] = `
"Association Association1
  documentation
  \\"doc\\"
  domain entity Domain1
    documentation
    \\"doc\\"
  domain entity Domain2
    documentation
    \\"doc\\"
  integer Property2
    documentation
    \\"doc\\"
    is part of identity
    max value 100
"
`;
